Qi/ catalog scatter linear regression

panchyo0 opened the pull request:
Add linear regression fitting function for catalog scatter plot
1. fitting will apply to all scatter data if no data points selected, double click on plots will remove the linear regression
2. display the linear function, variance-covariance matrix(cov00/cov01/cov11) and the sum of squares of the residuals as rss at the bottom information area.
3. performance: About 70ms to do the calculation and plot for 1M data source (i5 3GHz, 16 GB memory)
4. add statistics (mean/min/max/rms/std/count) for user selected column in scatter or histogram plot
https://user-images.githubusercontent.com/22226157/140429146-a2bdc83b-4e2d-4213-b335-b3c3d73cb97e.mp4

panchyo0 committed:
simple linear regression for catalog scatter plot

panchyo0 committed:
apply linear fitting to all data

panchyo0 committed:
swith to scientific notation

panchyo0 committed:
calculate mean

panchyo0 committed:
reformat

panchyo0 committed:
minor bug fix

kswang1029 commented:
The linear fit function works fine. Few improvement to consider:
1) the "statistic" dropdown is not consistent with the one in other widgets such as spectral profile widget. Maybe we can call it "show statistics".
2) the presentation of the fitting results (showing as %.4e) could be improved by displaying them in a new line
3) instead of just showing mean statistic, I suggest we also compute and show other common ones such as min/max/rms/std etc. The presentation would be in a new line too.

2 and 3 might be improved by using a dedicated box but I am not sure where to put it properly. For 2, maybe we can show those info as an annotation on the plot directly at the top-left corner. 


![Screen Shot 2021-11-08 at 8 06 08 PM](https://user-images.githubusercontent.com/20819712/140739180-442814e7-3068-44d0-9dd1-1d228c2b5e28.png)

4) for the "statistic" dropdown in the histogram plot widget, I am not sure if we need that. It would be good to also call it as "show statistics" as a _toggle_ and once it is enabled, we show the statistics of the column being used to generate the histogram.

panchyo0 committed:
show linear info as annotation at top left; add std, min, count, max and rms for statistics

panchyo0 commented:
@kswang1029  I updated the linear info as an annotation on the top left corner, Also added min/max/rms/std/count for statistics. Please check. 
> for the "statistic" dropdown in the histogram plot widget, I am not sure if we need that. It would be good to also call it as "show statistics" as a toggle and once it is enabled, we show the statistics of the column being used to generate the histogram.

I renamed this dropdown to `Statistic Source`. `can plot one column (DEC) and then show the statistic for another column (flux…).` is a feature request. So we might need to keep the dropdown 🤔

panchyo0 committed:
minor issue fix

panchyo0 committed:
Merge branch 'dev' into qi/issue_1240

veggiesaurus commented:
@kswang1029 please re-review

kswang1029 commented:
@panchyo0 it appears that if there are NaN data in a column, linear fit does not work, neither the statistics.

![Screen Shot 2021-11-16 at 10 55 36 AM](https://user-images.githubusercontent.com/20819712/141887720-c5e946aa-651e-4f42-a426-b0f0a280d776.png)

panchyo0 committed:
replace nan with 0

panchyo0 committed:
reformat

panchyo0 committed:
add valid count

panchyo0 committed:
reformat

panchyo0 committed:
fix  valid count bug

kswang1029 reviewed (approved):
This now works nicely. One last change is to display numbers as %.4e instead of the current %.2e for better UX. This applies to the displayed fitting results and statistics.

panchyo0 committed:
add sqrt cov00, sqrt cov11. update to %.4e

YuHsuan-Hwang commented on src/components/CatalogOverlay/CatalogPlotComponent/CatalogPlotComponent.tsx:
```diff
@@ -611,6 +703,25 @@ export class CatalogPlotComponent extends React.Component<WidgetProps> {
         );
 
+        const renderStatisticSelect = (
+            <FormGroup inline={true} label="Statistic Source">
```
Statistic "Source" is hidden in the histogram panel. We could either set a larger default width or try to adjust the column widths
![image](https://user-images.githubusercontent.com/43841102/142819128-1a0848c9-86db-4be8-884f-8a045a1bfa52.png)

YuHsuan-Hwang commented on src/components/CatalogOverlay/CatalogPlotComponent/CatalogPlotComponent.tsx:
```diff
@@ -702,13 +813,53 @@ export class CatalogPlotComponent extends React.Component<WidgetProps> {
+                    y1: fitting.intercept + fitting.slope * minMaxX.maxVal,
+                    line: {
+                        color: Colors.GREEN2,
+                        width: 5
```
change to `2.5 * ratio` so that the line width looks the same on a devicePixelRatio = 1 screen
![image](https://user-images.githubusercontent.com/43841102/142820553-9773b6d5-2570-4e68-befe-d4fd6a715ff6.png)

YuHsuan-Hwang commented on src/components/CatalogOverlay/CatalogPlotComponent/CatalogPlotComponent.tsx:
```diff
@@ -702,13 +813,53 @@ export class CatalogPlotComponent extends React.Component<WidgetProps> {
+                    text: widgetStore.fittingResultString,
+                    showarrow: false,
+                    font: {
+                        size: 18,
```
same comment: `9 * ratio`

panchyo0 replied:
Thanks 👍

panchyo0 committed:
adjust width and font size
